src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java

```diff
@@ -0,0 +1,126 @@
+ */
+
+package hudson.plugins.build_timeout.operations;
+import static hudson.util.TimeUnit2.MILLISECONDS;
+import static hudson.util.TimeUnit2.MINUTES;
+
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import hudson.Extension;
+import hudson.model.AbstractBuild;
+import hudson.model.AbstractProject;
+import hudson.model.Executor;
+import hudson.model.Result;
+import hudson.model.BuildListener;
+import hudson.plugins.build_timeout.BuildTimeOutOperation;
+import hudson.plugins.build_timeout.BuildTimeOutOperationDescriptor;
+
+import com.chikli.hudson.plugin.naginator.FixedDelay;
+
+/**
+ * Abort the build.
+ */
+public class AbortAndRestartOperation extends BuildTimeOutOperation {
+    
+    private final String maxRestarts;
+    
+    private static final Logger log = Logger.getLogger(AbortAndRestartOperation.class.getName());
+    
+    /**
+     * @return max restarts.
+     */
+    public String getMaxRestarts() {
+        return maxRestarts;
+    }
+        
+    @DataBoundConstructor
+    public AbortAndRestartOperation(String maxRestarts){
+        this.maxRestarts = maxRestarts;
+    }
+    
+    private static boolean isPresent() {
+        try {
+            Class.forName("com.chikli.hudson.plugin.naginator.NaginatorScheduleAction");
+            return true;
+        } catch (ClassNotFoundException ex) {
+            log.log(Level.WARNING, "Naginator not available. ", ex);
+            return false;
+        }
+    }
+       
+    /**
+     * @param build
+     * @param listener
+     * @param effectiveTimeout
+     * @return
+     * @see hudson.plugins.build_timeout.BuildTimeOutOperation#perform(hudson.model.AbstractBuild, hudson.model.BuildListener, long)
+     */
+    @Override
+    public boolean perform(AbstractBuild<?, ?> build, BuildListener listener, long effectiveTimeout) {
+              
+        long effectiveTimeoutMinutes = MINUTES.convert(effectiveTimeout,MILLISECONDS);
+        // Use messages in hudson.plugins.build_timeout.Messages for historical reason.
+        listener.getLogger().println(hudson.plugins.build_timeout.Messages.Timeout_Message(
+                effectiveTimeoutMinutes,
+                hudson.plugins.build_timeout.Messages.Timeout_Aborted())
+        );
+        
+        Executor e = build.getExecutor();
+        if (e != null) {
+            e.interrupt(Result.ABORTED);
+        }
+        
+        if(isPresent()){
+            FixedDelay sd = new FixedDelay(0); //Reschedule now!
+            int maxRestarts = 0;
+            try {
+                maxRestarts = Integer.parseInt(build.getEnvironment(listener).expand(this.maxRestarts));
+                build.addAction(new com.chikli.hudson.plugin.naginator.NaginatorScheduleAction(maxRestarts, sd, false));
+            } catch (IOException e1) {
+                log.log(Level.WARNING, "Failed to expand environment variables. ", e1);

```

Comment:
It's useful for users to output logs to the console log of the build.
`listener.getLogger()` provides `PrintStream` for that.